src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,605 @@
+/***************************************************************************
+                         qgsmeshtracerenderer.cpp
+                         -------------------------
+    begin                : November 2019
+    copyright            : (C) 2019 by Vincent Cloarec
+    email                : vcloarec at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmeshtracerenderer.h"
+
+
+
+QgsVector QgsMeshVectorValueInterpolatorFromVertex::value( const QgsPointXY &point )
+{
+  updateFace( point );
+
+  if ( mCacheFaceIndex == -1 || mCacheFaceIndex >= mTriangularMesh.triangles().count() )
+    return ( QgsVector( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ) );
+
+  if ( ! mActiveFaceFlagValue.active( mTriangularMesh.trianglesToNativeFaces()[mCacheFaceIndex] ) )
+    return ( QgsVector( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ) );
+
+  face = mTriangularMesh.triangles().at( mCacheFaceIndex );
+
+
+  QgsPoint p1 = mTriangularMesh.vertices().at( face[0] );
+  QgsPoint p2 = mTriangularMesh.vertices().at( face[1] );
+  QgsPoint p3 = mTriangularMesh.vertices().at( face[2] );
+
+  QgsVector v1 = QgsVector( mDatasetValues.value( face[0] ).x(),
+                            mDatasetValues.value( face[0] ).y() );
+
+  QgsVector v2 = QgsVector( mDatasetValues.value( face[1] ).x(),
+                            mDatasetValues.value( face[1] ).y() );
+
+  QgsVector v3 = QgsVector( mDatasetValues.value( face[2] ).x(),
+                            mDatasetValues.value( face[2] ).y() );
+
+
+  return QgsMeshLayerUtils::interpolateVectorFromVerticesData(
+           p1,
+           p2,
+           p3,
+           v1,
+           v2,
+           v3,
+           point );
+}
+
+QgsMeshVectorValueInterpolator::QgsMeshVectorValueInterpolator( QgsTriangularMesh triangularMesh,
+    QgsMeshDataBlock datasetVectorValues,
+    QgsMeshDataBlock scalarActiveFaceFlagValues ):
+  mTriangularMesh( triangularMesh ),
+  mDatasetValues( datasetVectorValues ),
+  mActiveFaceFlagValue( scalarActiveFaceFlagValues )
+{
+
+}
+
+QgsMeshVectorValueInterpolator::~QgsMeshVectorValueInterpolator() {}
+
+void QgsMeshVectorValueInterpolator::updateFace( const QgsPointXY &point )
```

Comment:
I am not sure if you need to first check if the point is in triangle. Isn't it enough to just check if `interpolateVectorFromVerticesData` is not NaN?

Reply:
Yes, you are maybe right! But it will depend which of `interpolateVectorFromVerticesData` or `isInTriangleFace` is faster. If `interpolateVectorFromVerticesData` do the job as fast, no need to have the `isInTriangleFace` function.

Reply:
that how it is done elsewhere

1. from spatial index on triangular mesh find just triangles that are in your current bounding box/extent
2. on those selected triangles do `interpolateVectorFromVerticesData` to find if inside/ouside na data interpolated